c/driver: Inconsistent Error Types for adbc_get_table_schema

When querying `adbc_get_table_schema` for a table that does not exist, the different drivers are returning different exception types. Here is what postgres looks like:

```python
>>> uri = "postgresql://postgres:postgres@localhost:5432/pandas"
>>> with dbapi.connect(uri) as conn:
...         conn.adbc_get_table_schema("does_not_exist")

../../mambaforge/envs/pandas-dev/lib/python3.10/site-packages/adbc_driver_manager/dbapi.py:453: in adbc_get_table_schema
    handle = self._conn.get_table_schema(
adbc_driver_manager/_lib.pyx:769: in adbc_driver_manager._lib.AdbcConnection.get_table_schema
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   adbc_driver_manager._lib.OperationalError: ADBC_STATUS_IO (10): [libpq] Failed to execute query: ERROR:  relation "iris" does not exist
E   CONTEXT:  unnamed portal parameter $1 = '...'

adbc_driver_manager/_lib.pyx:385: OperationalError
```

By contrast here is what sqlite throws:

```
>>> with dbapi.connect("file:/tmp/4fd67ccb-203f-4ade-bb22-7dad59904ea9") as conn:
...         conn.adbc_get_table_schema("does_not_exist")

../../mambaforge/envs/pandas-dev/lib/python3.10/site-packages/adbc_driver_manager/dbapi.py:453: in adbc_get_table_schema
    handle = self._conn.get_table_schema(
adbc_driver_manager/_lib.pyx:769: in adbc_driver_manager._lib.AdbcConnection.get_table_schema
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   adbc_driver_manager._lib.InternalError: ADBC_STATUS_INTERNAL (9): [SQLite] Failed to prepare query: no such table: iris

adbc_driver_manager/_lib.pyx:385: InternalError
```

Would be nice to have those consistent

--- a/c/driver/postgresql/connection.cc
+++ b/c/driver/postgresql/connection.cc
@@ -1313,6 +1313,7 @@ AdbcStatusCode PostgresConnection::GetTableSchema(const char* catalog,
     if (std::string(error->sqlstate, 5) == "42P01") {
       return ADBC_STATUS_NOT_FOUND;
     }
+    return result;
   }
 
   auto uschema = nanoarrow::UniqueSchema();

--- a/c/driver/postgresql/connection.cc
+++ b/c/driver/postgresql/connection.cc
@@ -1308,7 +1308,12 @@ AdbcStatusCode PostgresConnection::GetTableSchema(const char* catalog,
   StringBuilderReset(&query);
 
   RAISE_ADBC(result_helper.Prepare());
-  RAISE_ADBC(result_helper.Execute());
+  auto result = result_helper.Execute();
+  if (result != ADBC_STATUS_OK) {
+    if (std::string(error->sqlstate, 5) == "42P01") {
+      return ADBC_STATUS_NOT_FOUND;
+    }
+  }
 
   auto uschema = nanoarrow::UniqueSchema();
   ArrowSchemaInit(uschema.get());

--- a/c/driver/postgresql/statement.cc
+++ b/c/driver/postgresql/statement.cc
@@ -651,6 +651,7 @@ int TupleReader::GetNext(struct ArrowArray* out) {
   struct ArrowArray tmp;
   NANOARROW_RETURN_NOT_OK(BuildOutput(&tmp, &error));
 
+  PQclear(result_);
   // Check the server-side response
   result_ = PQgetResult(conn_);
   const ExecStatusType pq_status = PQresultStatus(result_);

--- a/c/driver/sqlite/sqlite.c
+++ b/c/driver/sqlite/sqlite.c
@@ -877,6 +877,7 @@ AdbcStatusCode SqliteConnectionGetTableSchema(struct AdbcConnection* connection,
     return ADBC_STATUS_INTERNAL;
   }
 
+  // TODO(apache/arrow-adbc#1025): escape
   if (StringBuilderAppend(&query, "%s%s", "SELECT * FROM ", table_name) != 0) {
     StringBuilderReset(&query);
     SetError(error, "[SQLite] Call to StringBuilderAppend failed");
@@ -888,8 +889,8 @@ AdbcStatusCode SqliteConnectionGetTableSchema(struct AdbcConnection* connection,
       sqlite3_prepare_v2(conn->conn, query.buffer, query.size, &stmt, /*pzTail=*/NULL);
   StringBuilderReset(&query);
   if (rc != SQLITE_OK) {
-    SetError(error, "[SQLite] Failed to prepare query: %s", sqlite3_errmsg(conn->conn));
-    return ADBC_STATUS_INTERNAL;
+    SetError(error, "[SQLite] GetTableSchema: %s", sqlite3_errmsg(conn->conn));
+    return ADBC_STATUS_NOT_FOUND;
   }
 
   struct ArrowArrayStream stream = {0};

--- a/c/integration/duckdb/duckdb_test.cc
+++ b/c/integration/duckdb/duckdb_test.cc
@@ -75,6 +75,7 @@ class DuckDbConnectionTest : public ::testing::Test,
 
   void TestAutocommitDefault() { GTEST_SKIP(); }
   void TestMetadataGetTableSchema() { GTEST_SKIP(); }
+  void TestMetadataGetTableSchemaNotFound() { GTEST_SKIP(); }
   void TestMetadataGetTableTypes() { GTEST_SKIP(); }
 
  protected:

--- a/c/validation/adbc_validation.cc
+++ b/c/validation/adbc_validation.cc
@@ -430,6 +430,19 @@ void ConnectionTest::TestMetadataGetTableSchema() {
                                     {"strings", NANOARROW_TYPE_STRING, NULLABLE}}));
 }
 
+void ConnectionTest::TestMetadataGetTableSchemaNotFound() {
+  ASSERT_THAT(AdbcConnectionNew(&connection, &error), IsOkStatus(&error));
+  ASSERT_THAT(AdbcConnectionInit(&connection, &database, &error), IsOkStatus(&error));
+  ASSERT_THAT(quirks()->DropTable(&connection, "thistabledoesnotexist", &error),
+              IsOkStatus(&error));
+
+  Handle<ArrowSchema> schema;
+  ASSERT_THAT(AdbcConnectionGetTableSchema(&connection, /*catalog=*/nullptr,
+                                           /*db_schema=*/nullptr, "thistabledoesnotexist",
+                                           &schema.value, &error),
+              IsStatus(ADBC_STATUS_NOT_FOUND, &error));
+}
+
 void ConnectionTest::TestMetadataGetTableTypes() {
   ASSERT_THAT(AdbcConnectionNew(&connection, &error), IsOkStatus(&error));
   ASSERT_THAT(AdbcConnectionInit(&connection, &database, &error), IsOkStatus(&error));

--- a/c/validation/adbc_validation.h
+++ b/c/validation/adbc_validation.h
@@ -188,6 +188,7 @@ class ConnectionTest {
 
   void TestMetadataGetInfo();
   void TestMetadataGetTableSchema();
+  void TestMetadataGetTableSchemaNotFound();
   void TestMetadataGetTableTypes();
 
   void TestMetadataGetObjectsCatalogs();
@@ -219,6 +220,9 @@ class ConnectionTest {
   TEST_F(FIXTURE, MetadataCurrentDbSchema) { TestMetadataCurrentDbSchema(); }           \
   TEST_F(FIXTURE, MetadataGetInfo) { TestMetadataGetInfo(); }                           \
   TEST_F(FIXTURE, MetadataGetTableSchema) { TestMetadataGetTableSchema(); }             \
+  TEST_F(FIXTURE, MetadataGetTableSchemaNotFound) {                                     \
+    TestMetadataGetTableSchemaNotFound();                                               \
+  }                                                                                     \
   TEST_F(FIXTURE, MetadataGetTableTypes) { TestMetadataGetTableTypes(); }               \
   TEST_F(FIXTURE, MetadataGetObjectsCatalogs) { TestMetadataGetObjectsCatalogs(); }     \
   TEST_F(FIXTURE, MetadataGetObjectsDbSchemas) { TestMetadataGetObjectsDbSchemas(); }   \

--- a/go/adbc/driver/snowflake/driver.go
+++ b/go/adbc/driver/snowflake/driver.go
@@ -163,6 +163,10 @@ func errToAdbcErr(code adbc.Status, err error) error {
 		var sqlstate [5]byte
 		copy(sqlstate[:], []byte(sferr.SQLState))
 
+		if sferr.SQLState == "42S02" {
+			code = adbc.StatusNotFound
+		}
+
 		return adbc.Error{
 			Code:       code,
 			Msg:        sferr.Error(),

--- a/python/adbc_driver_postgresql/tests/test_lowlevel.py
+++ b/python/adbc_driver_postgresql/tests/test_lowlevel.py
@@ -33,6 +33,11 @@ def postgres(
             yield conn
 
 
+def test_connection_get_table_schema(postgres: adbc_driver_manager.AdbcConnection):
+    with pytest.raises(adbc_driver_manager.ProgrammingError, match="NOT_FOUND"):
+        postgres.get_table_schema(None, None, "thistabledoesnotexist")
+
+
 def test_query_trivial(postgres: adbc_driver_manager.AdbcConnection) -> None:
     with adbc_driver_manager.AdbcStatement(postgres) as stmt:
         stmt.set_sql_query("SELECT 1")

--- a/python/adbc_driver_sqlite/tests/test_lowlevel.py
+++ b/python/adbc_driver_sqlite/tests/test_lowlevel.py
@@ -29,6 +29,11 @@ def sqlite():
             yield conn
 
 
+def test_connection_get_table_schema(sqlite):
+    with pytest.raises(adbc_driver_manager.ProgrammingError, match="NOT_FOUND"):
+        sqlite.get_table_schema(None, None, "thistabledoesnotexist")
+
+
 def test_query_trivial(sqlite):
     with adbc_driver_manager.AdbcStatement(sqlite) as stmt:
         stmt.set_sql_query("SELECT 1")

